### PR TITLE
fix(treesitter): prevent endless loop on self-inheritence

### LINKED ISSUE #20139: nvim crashes occasionally and outputs errors always when using custom queries/x/highlights.scm with `inherits` modeline
### Neovim version (nvim -v)

NVIM v0.8.0-dev-1059-g2d6735d8c

### Vim (not Nvim) behaves the same?

wouldnt know

### Operating system/version

archlinux

### Terminal name/version

st 0.8.5

### $TERM environment variable

tmux-256color

### Installation

https://github.com/neovim/neovim/releases

### How to reproduce the issue

* have a nvim nightly from release page
* prepare ~/.local/share/nvim/treesitter/{query,query-info} -> {lua.so, lua.reversion}
* echo 'set rtp+=~/.local/share/nvim/treesitter' > ~/.config/nvim/init.vim
* create custom highlights.scm
* open a lua file
* `:lua vim.treesitter.start()`
* move arround

the highlights.cm i used `~/.config/nvim/after/queries/lua/highlights.scm`
```
;; inherits: lua

;; changes:
;; * treat end as operator

[
 "goto"
 "in"
] @keyword

[
 "local"
] @type.qualifier
```

### Expected behavior

if there are errors in my custom highlights.scm, nvim can give reasonable errors.

### Actual behavior

nvim will pour errors in the current buffer, seems like as a virtual text.

some errors (they occur arbitrarily to me):
```
treesitter/highlighter: Error executing lua: error in error handling
...d6735d8c/share/nvim/runtime/lua/vim/treesitter/query.lua:81: stack overflowting
 treesitter/highlighter: Error executing lua: vim/shared.lua:0: stack overflow
```
136a1b8c-309d-11ed-869c-54b203702c9c

### PR DESCRIPTION
**Pull request by vigoux**
@haolian9 could you try this PR and see if it fixes the issue on your
end ?

You should be getting an error now, and no crash.

Fixes #20139

### CODE REVIEW

**Comment by haolian9** [comment]
> You should be getting an error now, and no crash.

yes, i do. errors show in each line like this `treesitter/highlighter: Error executing lua: ...d6735d8c/share/nvim/runtime/lua/vim/treesitter/query.lua:41: Self-inheritance is prohibited: lua`
(i did not re-compile nvim, just replaced the query.lua runtime file)

**Comment by vigoux** [comment]
Sounds like it fixed the issue then.
I wonder if we should instead consider this case a `extends` @clason ?

That would avoid the error and behave better?

**Comment by clason** [comment]
Yes, this sounds like it should have used `extends` instead. But this is a documentation issue; I don't think we should silently change the meaning of directives?

**Comment by vigoux** [comment]
I think that, in this case, this makes the behavior of `inherits` more previsible ?
I'll add a documentation note anyway.

File: runtime/doc/treesitter.txt [hunk 1]
```diff
@@ -199,16 +199,20 @@ and predicates. A `capture` allows you to associate names with a specific
...
 "modelines", that is comments in the queries starting with `;`. Here are the
 currently supported modeline alternatives:
 
-    `inherits: {lang}...`
+    `inherits: {lang}...`                                 *ts-modeline-inherits*
         Specifies that this query should inherit the queries from {lang}.
         This will recursively descend in the queries of {lang} unless wrapped
         in parentheses: `({lang})`.
+        Note: this is meant to be used to include queries from another
+        language. If you want your query to extend the queries of this
```
Comment (clason) [inline runtime/doc/treesitter.txt:212]:
```suggestion
        language. If you want your query to extend the queries of the same
```

**Review by clason — APPROVED** [review]

**Pull request merged**